File: tutorial/go2_tutorial/src/go2_nav_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "go2_tutorial_inter/action/nav.hpp"
#include "nav_msgs/msg/odometry.hpp"
#include "geometry_msgs/msg/point.hpp"
#include "sport_model.hpp"


using namespace std::placeholders;
using namespace rclcpp_action;
using namespace std::chrono_literals;


// Go2NavServer 節點：提供導航功能的行動伺服器
class Go2NavServer : public rclcpp::Node{
public:
    // 構造函數，初始化節點名稱並建立伺服器與訂閱器
    Go2NavServer(): Node("go2_nav_server"){
        RCLCPP_INFO(this->get_logger(), "Go2NavServer created!"); // 節點建立提示
        
        // 建立參數 client，連線到 go2_ctrl 節點
        param_client_ = std::make_shared<rclcpp::AsyncParametersClient>(this, "go2_ctrl");
        while (!param_client_->wait_for_service(1s)){
            if (!rclcpp::ok()){
                return; // 如果 ROS2 系統已關閉，結束等待
            }
            RCLCPP_INFO(this->get_logger(), "loading...."); // 等待參數服務啟動
        }

        // 建立導航行動伺服器
        nav_server_ = rclcpp_action::create_server<go2_tutorial_inter::action::Nav>(
            this,
            "nav",
            std::bind(&Go2NavServer::goal_cb, this, _1, _2), // 目標回呼函式
            std::bind(&Go2NavServer::cancel_cb, this, _1), // 取消回呼函式
            std::bind(&Go2NavServer::accepted_cb, this, _1) // 接受回呼函式
        );

        // 訂閱 odom topic，取得機器人目前位置
        odom_sub_ = this->create_subscription<nav_msgs::msg::Odometry>(
            "odom", 10, std::bind(&Go2NavServer::odom_cb, this, _1)
        );

        // 宣告參數，包含導航目標誤差範圍
        this->declare_parameter("x", 0.1);
        this->declare_parameter("error", 0.2);

    }
private:
    rclcpp_action::Server<go2_tutorial_inter::action::Nav>::SharedPtr nav_server_; // 導航伺服器
    rclcpp::AsyncParametersClient::SharedPtr param_client_; // 參數 client

    geometry_msgs::msg::Point current_point_, start_point_; // 目前位置與起始位置

    // 目標回呼函式，處理導航目標請求
    GoalResponse goal_cb(const GoalUUID &uuid, std::shared_ptr<const go2_tutorial_inter::action::Nav::Goal> goal){
        float goal_dis = goal->goal;
        (void)uuid; 
        if (goal_dis > 0.0){
            start_point_ = current_point_; // 設定起始位置
            RCLCPP_INFO(this->get_logger(), "Goal accepted");
            param_client_->set_parameters({
                rclcpp::Parameter("sport_api_id", ROBOT_SPORT_API_MOVE),
                this->get_parameter("x")
       
            });

            
            return GoalResponse::ACCEPT_AND_EXECUTE; // 接受目標
        }
        else{
            RCLCPP_INFO(this->get_logger(), "Goal rejected");
            return GoalResponse::REJECT; // 拒絕目標
        }

    };

    // 取消回呼函式，處理取消導航請求
    CancelResponse cancel_cb(std::shared_ptr<ServerGoalHandle<go2_tutorial_inter::action::Nav>> server_go_handler){
        (void)server_go_handler;
        RCLCPP_INFO(this->get_logger(), "Goal cancel request received");
        stop_move(); // 停止移動
        
        return CancelResponse::ACCEPT; // 接受取消請求
    };

    // 停止移動的方法
    void stop_move(){
        param_client_->set_parameters({
            rclcpp::Parameter("sport_api_id", ROBOT_SPORT_API_STOPMOVE),
            rclcpp::Parameter("x", 0.0),
            rclcpp::Parameter("y", 0.0),
            rclcpp::Parameter("z", 0.0)
        });
    }

    // 接受回呼函式，處理接受的導航請求
    void accepted_cb(std::shared_ptr<ServerGoalHandle<go2_tutorial_inter::action::Nav>> server_goal_handle){
        std::thread(std::bind(&Go2NavServer::execute, this, _1), server_goal_handle).detach();

    }

    // 執行導航的主要邏輯
    void execute(std::shared_ptr<ServerGoalHandle<go2_tutorial_inter::action::Nav>> server_goal_handle){
        rclcpp::Rate rate(1.0); // 設定執行頻率
        auto feedback = std::make_shared<go2_tutorial_inter::action::Nav::Feedback>();
        auto nav_result = std::make_shared<go2_tutorial_inter::action::Nav::Result>();
        while(rclcpp::ok()){

            auto dis_x = current_point_.x - start_point_.x;
            auto dis_y = current_point_.y - start_point_.y;
            auto dis = sqrt(dis_x*dis_x + dis_y*dis_y); // 計算距離

            auto distance = server_goal_handle->get_goal()->goal - dis;

            feedback->distance = distance; // 更新回饋距離
            server_goal_handle->publish_feedback(feedback); // 發布回饋

            if (server_goal_handle->is_canceling()){ // 如果取消請求
                stop_move();
                nav_result->point = current_point_;
                server_goal_handle->canceled(nav_result);
                return;
            }

            if (distance <= this->get_parameter("error").as_double()){ // 如果到達目標
                RCLCPP_INFO(this->get_logger(), "Goal reached");
                stop_move();
                break;
            }
            rate.sleep();
            
            
        
        
        }
        if (rclcpp::ok()){
                stop_move();
                nav_result->point = current_point_;
                server_goal_handle->succeed(nav_result); // 導航成功
        }

    }


    rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_sub_; // odometry 訂閱器

    // odometry 回呼函式，更新目前位置
    void odom_cb(const nav_msgs::msg::Odometry::SharedPtr odom){
        current_point_ = odom->pose.pose.position;
    }   


};



int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv); // 初始化 ROS2 系統
  rclcpp::spin(std::make_shared<Go2NavServer>()); // 執行 Go2Ctrl 節點，持續運作直到結束
  rclcpp::shutdown(); // 關閉 ROS2 系統
  return 0;
}

File: tutorial/go2_tutorial_py/go2_tutorial_py/go2_nav_server.py

import rclpy
from rclpy.node import Node
from rclpy.action import ActionServer
from rclpy.action.server import CancelResponse, GoalResponse, ServerGoalHandle
from go2_tutorial_inter.action import Nav
from rclpy.executors import MultiThreadedExecutor
from geometry_msgs.msg import Point
from nav_msgs.msg import Odometry
from unitree_api.msg import Request
from .sport_model import ROBOT_SPORT_API_IDS
import json
import time
import math


class Go2NavServer(Node):
    # 初始化函式，設定 ROS2 節點、參數、訂閱與發布
    def __init__(self):
        super().__init__("go2_nav_server")
        self.get_logger().info("Go2 Control Node has been started.")  # 節點啟動訊息

        self.point = Point()  # 儲存當前位置

        # 訂閱 odom 主題以獲取機器人位置
        self.odom_sub = self.create_subscription(Odometry, "odom", self.odom_cb, 10)
        # 宣告三個參數 x, y, z
        self.declare_parameter("x", 0.1)
        self.declare_parameter("y", 0.0)
        self.declare_parameter("z", 0.0)

        # 建立請求的 publisher
        self.req_pub = self.create_publisher(Request, "/api/sport/request", 10)
        # 設定定時器，週期性發送請求
        self.timer = self.create_timer(0.1, self.on_timer)
        # 初始 API 狀態為平衡站立
        self.api_id = ROBOT_SPORT_API_IDS["BALANCESTAND"]

        # 建立 action server，負責導航任務
        self.action_server = ActionServer(
            self,
            Nav,
            'nav',
            self.execute,
            goal_callback=self.goal_callback,
            cancel_callback=self.cancel_callback
        )

    # action server 的執行函式，負責處理導航目標
    def execute(self, goal_handle : ServerGoalHandle):
        feedback = Nav.Feedback()  # 建立回饋物件
        while rclpy.ok():
            time.sleep(0.5)  # 每 0.5 秒檢查一次
            # 計算目前與起始點的距離
            dis_x = self.point.x - self.start_point.x
            dis_y = self.point.y - self.start_point.y
            dis = math.sqrt(math.pow(dis_x, 2) + math.pow(dis_y, 2))
            feedback.distance = goal_handle.request.goal - dis  # 計算剩餘距離
            goal_handle.publish_feedback(feedback)  # 發布回饋
            if feedback.distance <= 0.0:
                self.api_id = ROBOT_SPORT_API_IDS["STOPMOVE"]  # 到達目標則停止移動
                break
        goal_handle.succeed()  # 任務成功
        result = Nav.Result()
        result.point = self.point  # 回傳最終位置
        return result
        

    # 接收 action goal 的 callback，檢查目標是否合法
    def goal_callback(self, goal_request : Nav.Goal):
        if goal_request.goal > 0.0:
            self.start_point = self.point  # 記錄起始點
            self.get_logger().info(f"Received goal: {goal_request.goal}")
            self.api_id = ROBOT_SPORT_API_IDS["MOVE"]  # 設定為移動模式
            return GoalResponse.ACCEPT  # 接受目標
        else:
            self.api_id = ROBOT_SPORT_API_IDS["STOPMOVE"]  # 目標不合法則停止
            self.get_logger().error("Goal must be positive.")
            return GoalResponse.REJECT  # 拒絕目標


    # action 取消時的 callback，收到取消請求時停止移動
    def cancel_callback(self, goal_handle):
        self.api_id = ROBOT_SPORT_API_IDS["STOPMOVE"]  # 設定為停止
        return CancelResponse.ACCEPT


    # odom 訂閱回調，更新目前位置
    def odom_cb(self, odom : Odometry):
        self.point = odom.pose.pose.position

    # 定時器回調，週期性發送控制請求到下游
    def on_timer(self):
        req = Request()
        req.header.identity.api_id = self.api_id  # 設定 API 狀態
        js = {
            "x": self.get_parameter("x").value,
            "y": self.get_parameter("y").value,
            "z": self.get_parameter("z").value
        }
        req.parameter = json.dumps(js)  # 將參數轉為 JSON 字串
        self.req_pub.publish(req)  # 發布請求


# 主程式進入點，啟動 ROS2 節點
def main():
    rclpy.init()  # 初始化 ROS2
    node = Go2NavServer()  # 建立導航伺服器節點
    excutor = MultiThreadedExecutor()  # 多執行緒執行器
    excutor.add_node(node)
    excutor.spin()  # 開始執行
    rclpy.shutdown()  # 關閉 ROS2
if __name__ == "__main__":
    main()
    